src/Cardano/Wallet/AddressDerivation.hs

```diff
@@ -0,0 +1,304 @@
+-- used for addresses that are part of the 'advertised' targets of a transaction
+data ChangeChain
+    = InternalChain
+    | ExternalChain
+    deriving (Generic, Show, Eq)
+
+instance NFData ChangeChain
+
+-- Not deriving 'Enum' because this could have a dramatic impact if we were
+-- to assign the wrong index to the corresponding constructor (by swapping
+-- around the constructor above for instance).
+instance Enum ChangeChain where
+    toEnum = \case
+        0 -> ExternalChain
+        1 -> InternalChain
+        _ -> error "ChangeChain.toEnum: bad argument"
+    fromEnum = \case
+        ExternalChain -> 0
+        InternalChain -> 1
+
+-- | Purpose is a constant set to 44' (or 0x8000002C) following the BIP-44
+-- recommendation. It indicates that the subtree of this node is used
+-- according to this specification.
+--
+-- Hardened derivation is used at this level.
+purposeIndex :: Word32
+purposeIndex = 0x8000002C
+
+-- | One master node (seed) can be used for unlimited number of independent
+-- cryptocoins such as Bitcoin, Litecoin or Namecoin. However, sharing the
+-- same space for various cryptocoins has some disadvantages.
+--
+-- This level creates a separate subtree for every cryptocoin, avoiding reusing
+-- addresses across cryptocoins and improving privacy issues.
+--
+-- Coin type is a constant, set for each cryptocoin. For Cardano this constant
+-- is set to 1815' (or 0x80000717). 1815 is the birthyear of our beloved Ada
+-- Lovelace.
+--
+-- Hardened derivation is used at this level.
+coinTypeIndex :: Word32
+coinTypeIndex = 0x80000717
+
+-- | Generate a new key from seed. Note that the @depth@ is left open so that
+-- the caller gets to decide what type of key this is. This is mostly for
+-- testing, in practice, seeds are used to represent root keys, and one should
+-- use 'generateKeyFromSeed'.
+unsafeGenerateKeyFromSeed
+    :: (ByteString, Passphrase "generation")
+        -- ^ The actual seed and its recovery / generation passphrase
+    -> Passphrase "encryption"
+    -> Key depth XPrv
+unsafeGenerateKeyFromSeed (seed, Passphrase recPwd) (Passphrase encPwd) =
+    Key $ generateNew seed recPwd encPwd
+
+-- | Generate a root key from a corresponding seed
+generateKeyFromSeed
+    :: (ByteString, Passphrase "generation")
+        -- ^ The actual seed and its recovery / generation passphrase
+    -> Passphrase "encryption"
+    -> Key 'RootK XPrv
+generateKeyFromSeed = unsafeGenerateKeyFromSeed
+
+-- | Derives account private key from the given root private key, using
+-- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+-- package for more details).
+--
+-- NOTE: The caller is expected to provide the corresponding passphrase.
+deriveAccountPrivateKey
+    :: Passphrase "encryption"
+    -> Key 'RootK XPrv
+    -> Index 'Hardened 'AccountK
+    -> Key 'AccountK XPrv
+deriveAccountPrivateKey (Passphrase pwd) (Key rootXPrv) (Index accIx) =
+    let
```

Comment:
I notice there was a passphrase check in the previous version.
This function is no longer failable (partly because the `isHardened` check is obviated by the `Index` type).
What happens when the key encryption passphrase is wrong?

Reply:
I left a comment about that but I'll extend it. This has been left to the caller as controlling the password would require to either keep a hash of the password in memory (which was a wrong design decision) like before, or have an access to the keystore / password vault. This is a responsibility of the wallet layer calling those derivation primitives.